### PR TITLE
Add CodeQL Analysis workflow

### PR DESCRIPTION
**Pull request by andy5995**

### CODE REVIEW

**Comment by andy5995** [comment]
It doesn't run. The message is

```
github/codeql-action/init@v3, github/codeql-action/autobuild@v3, and github/codeql-action/analyze@v3 are not allowed to be used in geany/geany. Actions in this workflow must be: within a repository owned by geany or matching the following: actions/cache@v2, actions/setup-python@v2, actions/cache@v3, actions/upload-artifact@v3, actions/checkout@v3, docker/login-action@v2, actions/setup-python@v4, actions/delete-package-versions@v4, msys2/setup-msys2@v2, actions/cache@v4, actions/checkout@v4, actions/upload-artifact@v4, docker/login-action@v3, jwalton/gh-find-current-pr@v1, actions/delete-package-versions@v5, actions/setup-python@v5.
```

**Comment by elextr** [comment]
Maybe you could try the [CLI](https://docs.github.com/en/code-security/codeql-cli) version and see if its useful first.

**Comment by andy5995** [comment]
I enabled the default scan from my repository settings. The only problems it found were related to Scintilla https://github.com/andy5995/geany/security/code-scanning

**Comment by elextr** [comment]
Erm, the link doesn't work.

**Comment by rdipardo** [comment]
> Erm, the link doesn't work.

That's because:

> [y]ou need write permission to view a summary of all the alerts for a repository on the **Security** tab.
>
> <https://docs.github.com/en/code-security/code-scanning/managing-code-scanning-alerts/managing-code-scanning-alerts-for-your-repository>

Only scans of _pull requests_ are publicly visible, *per* the docs.

**Comment by andy5995** [comment]
@rdipardo Thank you for the explanation. @elextr I sent you an invite to be a collaborator.

Another thing CodeQL does is when a workflow is in place, it gives alerts for any problematic code that may be introduced in a pull request.

**Comment by eht16** [comment]
Sounds good.

- For the next time, a description to describe what is the PR about and more importantly why, is generally a good idea.
- We could add `*.py` to the patterns and `python` to the matrix to also cover the Python helper scripts in the repository.
- Also, could you strip the commented code from the workflow definition and also the extensive description comments? I think they are not very useful.
- I enabled the mentioned "github/*" actions. In general we restrict the usable actions to the necessary ones for security reasons.

**Comment by eht16** [comment]
Adding `  workflow_dispatch:` might be a good idea to manually re-run the workflow.

**Comment by andy5995** [comment]
> Sounds good.
> 
>     * For the next time, a description to describe what is the PR about and more importantly why, is generally a good idea.

Understood!

> 
>     * We could add `*.py` to the patterns and `python` to the matrix to also cover the Python helper scripts in the repository.

Done.

> 
>     * Also, could you strip the commented code from the workflow definition and also the extensive description comments? I think they are not very useful.

Done.

>Adding   workflow_dispatch: might be a good idea to manually re-run the workflow.

Done. Some examples of this workflow include a cron schedule as well. Would you like that added, and if so, how often should it be scheduled?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -8,6 +8,13 @@ on:
   push:
     branches:
       - master
+    - '**'
```
Comment (eht16) [inline .github/workflows/build.yml:11]: Why did you change this, what's the intention?

**Comment by eht16** [comment]
Thanks!

> Done. Some examples of this workflow include a cron schedule as well. Would you like that added, and if so, how often should it be scheduled?

I would say once in a month is a good start.

File: .github/workflows/build.yml (hunk 1 above)
Comment (eht16) [reply .github/workflows/build.yml:11]: And it causes an error:
```
Invalid workflow file: .github/workflows/build.yml#L8
You have an error in your yaml syntax on line 8
```

Comment (andy5995) [reply .github/workflows/build.yml]: So it won't run when other yml files are changed. Trivial in this case, but I figured you might want to add more path filters in the future:

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

Comment (eht16) [reply .github/workflows/build.yml]: I don't see why, at least for this case it seems a bit beyond the goal.

If we have use cases for excluding files from triggering a CI run, we
can add them but no excluding only two YAML files doesn't change much.

Without having checked it, I think about 95% of the commits will change files which should trigger the CI anyway, so don't waste resources maintaining an articial exclude list.

**Comment by eht16** [comment]
> Adding ` workflow_dispatch:` might be a good idea to manually re-run the workflow.

I would still recommend this simple change.

And it seems the configuration still has syntax errors: https://github.com/geany/geany/actions/runs/8135027731/workflow

**Comment by andy5995** [comment]
> > Adding ` workflow_dispatch:` might be a good idea to manually re-run the workflow.
> 
> I would still recommend this simple change.
> 
> And it seems the configuration still has syntax errors: https://github.com/geany/geany/actions/runs/8135027731/workflow

That seems outdated. I added the workflow_dispatch and fixed the cron syntax/indentation about a week ago.

File: .github/workflows/build.yml (hunk 1 above)
Comment (andy5995) [reply .github/workflows/build.yml]: I totally see your point. I've reverted all changes to build.yml.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@ on:
 
 # cancel already running builds of the same branch or pull request
 concurrency:
-  group: ci-${{ github.head_ref }} || concat(${{ github.ref }}
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}
```
Comment (andy5995) [inline .github/workflows/build.yml:18]: I've changed this because the codeql job cancelled itself, considering the build job a higher priority

**Comment by eht16** [comment]
> > > Adding ` workflow_dispatch:` might be a good idea to manually re-run the workflow.
> > 
> > 
> > I would still recommend this simple change.
> > And it seems the configuration still has syntax errors: https://github.com/geany/geany/actions/runs/8135027731/workflow
> 
> That seems outdated. I added the workflow_dispatch and fixed the cron syntax/indentation about a week ago.

Sorry, I guess I didn't properly refresh this page and so commented on the old code.

Looks good to me now.
Thanks.

**Comment by andy5995** [comment]
No worries @eht16 , I know mistakes happen to humans. Cheers!
